dojo/tools/bugcrowd_api/parser.py

```diff
@@ -118,27 +129,31 @@ def convert_endpoint(self, url):
         if result:
             return Endpoint.from_uri(result.group(0))
         else:
-            return Endpoint.from_uri('')
+            return Endpoint.from_uri("")
```

Comment:
Not sure this is ok.
@kiblik any feedback?

Reply:
Hi @Gby56,
unfortunately, I have to agree with @damiencarol.

I checked the implementation of the parser but I couldn't find why your lines need to use regex for searching/validating URL and with using of `Endpoint.from_uri(...)`. Plus I'm also not sure what will even `Endpoint.from_uri("")` return (with high probability `ValidationError`).

In my opinion,
```python
            if bug_url != "" and bug_url is not None:
                endpoint = self.convert_endpoint(bug_url)
                finding.unsaved_endpoints = [endpoint]
```
could be replaced by
```python
            if bug_url:
                endpoint = Endpoint.from_uri(bug_url)
                finding.unsaved_endpoints = [endpoint]
```
and `convert_endpoint(...)` could be removed.
 
In your JSONs, I wasn't able to see any `bug_url` which needs this kind of "special care" :)

By the way, as I remember correctly, the URL should not be mentioned in the description since the `Endpoint` and `Endpoint_Status` work quite well.

Reply:
Hello ! Yes the regex seems a bit odd but I kept getting a bug on my own private submissions, which I can't share of course, but let me try to find an example and maybe the fix could be implemented in the endpoint from_uri method instead, it was a weird one

Reply:
And I agree I'll remove the empty endpoint for sure, that was creating a useless endpoint I think

Reply:
If URL is not valid, `Endpoint.from_uri(...)` will raise `ValidationError` so 
```python
            if bug_url:
                try:
                    endpoint = Endpoint.from_uri(bug_url)
                    finding.unsaved_endpoints = [endpoint]
```
should be good.

Reply:
100% agree. good feedback @kiblik

Reply:
The error I keep getting with the `Endpoint.from_uri` is 
```An exception error occurred during the report import:invalid scheme: ' http'. Only alphanumeric, "+", "-", and "." allowed. Did you meant to call URL.from_text()?```

Reply:
I'll try to ignore the exception but how can I return a message to the user in the UI, to tell them that a URL was not parsed, but not fail the whole parser ?

Reply:
It looks like there problem with space at the beginning of the URL. I see you used `strip()` in your implementation. Can you try to call `Endpoint.from_uri(bug_url.strip())`. It may fix the whole issue.

Reply:
I'll leave the logger because the parser is really going to get tested with a lot of weird inputs from submissions, so I think it's better not to fully break the import for 1 single submission

Reply:
Even if you add `strip()`, feel free to use `try ... except ...`. It is a good approach.

But setting `finding.unsaved_endpoints = []` is not necessary. It is `[]` by default.

Reply:
I keep seeing a weird error now though, Mysql doesn't seem too happy about a protocol ?
```
uwsgi_1         | Traceback (most recent call last):
uwsgi_1         |   File "/app/./dojo/engagement/views.py", line 626, in import_scan_results
uwsgi_1         |     test, finding_count, closed_finding_count, _ = importer.import_scan(scan, scan_type, engagement, user, environment, active=active, verified=verified, tags=tags,
uwsgi_1         |   File "/app/./dojo/importers/importer/importer.py", line 324, in import_scan
uwsgi_1         |     new_findings = self.process_parsed_findings(test, parsed_findings, scan_type, user, active,
uwsgi_1         |   File "/app/./dojo/decorators.py", line 47, in __wrapper__
uwsgi_1         |     return func(*args, **kwargs)
uwsgi_1         |   File "/usr/local/lib/python3.8/site-packages/celery/local.py", line 188, in __call__
uwsgi_1         |     return self._get_current_object()(*a, **kw)
uwsgi_1         |   File "/usr/local/lib/python3.8/site-packages/celery/app/task.py", line 392, in __call__
uwsgi_1         |     return self.run(*args, **kwargs)
uwsgi_1         |   File "/app/./dojo/importers/importer/importer.py", line 136, in process_parsed_findings
uwsgi_1         |     importer_utils.add_endpoints_to_unsaved_finding(item, test, item.unsaved_endpoints, sync=True)
uwsgi_1         |   File "/app/./dojo/decorators.py", line 47, in __wrapper__
uwsgi_1         |     return func(*args, **kwargs)
uwsgi_1         |   File "/usr/local/lib/python3.8/site-packages/celery/local.py", line 188, in __call__
uwsgi_1         |     return self._get_current_object()(*a, **kw)
uwsgi_1         |   File "/usr/local/lib/python3.8/site-packages/celery/app/task.py", line 392, in __call__
uwsgi_1         |     return self.run(*args, **kwargs)
uwsgi_1         |   File "/app/./dojo/importers/utils.py", line 135, in add_endpoints_to_unsaved_finding
uwsgi_1         |     ep, created = endpoint_get_or_create(
uwsgi_1         |   File "/app/./dojo/endpoint/utils.py", line 81, in endpoint_get_or_create
uwsgi_1         |     return Endpoint.objects.get_or_create(**kwargs)
uwsgi_1         |   File "/usr/local/lib/python3.8/site-packages/django/db/models/manager.py", line 85, in manager_method
uwsgi_1         |     return getattr(self.get_queryset(), name)(*args, **kwargs)
uwsgi_1         |   File "/usr/local/lib/python3.8/site-packages/tagulous/models/tagged.py", line 242, in get_or_create
uwsgi_1         |     return super(TaggedQuerySet, self).get_or_create(defaults, **safe_fields)
uwsgi_1         |   File "/usr/local/lib/python3.8/site-packages/django/db/models/query.py", line 588, in get_or_create
uwsgi_1         |     return self.create(**params), True
uwsgi_1         |   File "/usr/local/lib/python3.8/site-packages/tagulous/models/tagged.py", line 223, in create
uwsgi_1         |     obj = super(TaggedQuerySet, self).create(**safe_fields)
uwsgi_1         |   File "/usr/local/lib/python3.8/site-packages/django/db/models/query.py", line 453, in create
uwsgi_1         |     obj.save(force_insert=True, using=self.db)
uwsgi_1         |   File "/usr/local/lib/python3.8/site-packages/django/db/models/base.py", line 739, in save
uwsgi_1         |     self.save_base(using=using, force_insert=force_insert,
uwsgi_1         |   File "/usr/local/lib/python3.8/site-packages/django/db/models/base.py", line 776, in save_base
uwsgi_1         |     updated = self._save_table(
uwsgi_1         |   File "/usr/local/lib/python3.8/site-packages/django/db/models/base.py", line 881, in _save_table
uwsgi_1         |     results = self._do_insert(cls._base_manager, using, fields, returning_fields, raw)
uwsgi_1         |   File "/usr/local/lib/python3.8/site-packages/django/db/models/base.py", line 919, in _do_insert
uwsgi_1         |     return manager._insert(
uwsgi_1         |   File "/usr/local/lib/python3.8/site-packages/django/db/models/manager.py", line 85, in manager_method
uwsgi_1         |     return getattr(self.get_queryset(), name)(*args, **kwargs)
uwsgi_1         |   File "/usr/local/lib/python3.8/site-packages/django/db/models/query.py", line 1270, in _insert
uwsgi_1         |     return query.get_compiler(using=using).execute_sql(returning_fields)
uwsgi_1         |   File "/usr/local/lib/python3.8/site-packages/django/db/models/sql/compiler.py", line 1416, in execute_sql
uwsgi_1         |     cursor.execute(sql, params)
uwsgi_1         |   File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 66, in execute
uwsgi_1         |     return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
uwsgi_1         |   File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 75, in _execute_with_wrappers
uwsgi_1         |     return executor(sql, params, many, context)
uwsgi_1         |   File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
uwsgi_1         |     return self.cursor.execute(sql, params)
uwsgi_1         |   File "/usr/local/lib/python3.8/site-packages/django/db/utils.py", line 90, in __exit__
uwsgi_1         |     raise dj_exc_value.with_traceback(traceback) from exc_value
uwsgi_1         |   File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
uwsgi_1         |     return self.cursor.execute(sql, params)
uwsgi_1         |   File "/usr/local/lib/python3.8/site-packages/django/db/backends/mysql/base.py", line 73, in execute
uwsgi_1         |     return self.cursor.execute(query, args)
uwsgi_1         |   File "/usr/local/lib/python3.8/site-packages/MySQLdb/cursors.py", line 206, in execute
uwsgi_1         |     res = self._query(query)
uwsgi_1         |   File "/usr/local/lib/python3.8/site-packages/MySQLdb/cursors.py", line 319, in _query
uwsgi_1         |     db.query(q)
uwsgi_1         |   File "/usr/local/lib/python3.8/site-packages/MySQLdb/connections.py", line 254, in query
uwsgi_1         |     _mysql.connection.query(self, query)
uwsgi_1         | django.db.utils.DataError: (1406, "Data too long for column 'protocol' at row 1")
```

Reply:
I think the endpoint model in defectdojo is too restrictive

Reply:
- `Endpoint` doesn't accept only well-known protocols. It accepts also unknown protocols.
- `Endpoint` uses well-known protocols but only in `__str__` function - it can decide to hide the port (e.g. if the protocol is HTTP and the port is 80 or if the protocol is HTTPS and the port is 443)
- I fully support extending the length of the protocol in reasonable cases (there are 20 characters right now). Can tell us what is your protocol?
- We tried to avoid so obscure protocol names - please check https://github.com/DefectDojo/django-DefectDojo/pull/4696#discussion_r655532107
- `Endpoint` tried to reflect RFC definition for URI/URL. 
- We already implemented less strict validation even if it doesn't follow RFC. E.g. https://github.com/DefectDojo/django-DefectDojo/blob/master/unittests/test_endpoint_model.py#L68-L74. More info here https://github.com/DefectDojo/django-DefectDojo/pull/4604. But we would like to avoid very unusual formats only because some other developer in some other project doesn't follow RFCs.

Reply:
Gotcha ! That's good to know then, I guess it's mostly just the input that I send into from_uri that is creating issues, I need to isolate the URI/URL first and then try to create an Endpoint from that
I have tried to use that regex for validation of a URI no matter the protocol etc... but it's a lot of trouble for no reason, just to isolate a URI in a random string https://github.com/spring-projects/spring-framework/blob/main/spring-web/src/main/java/org/springframework/web/util/UriComponentsBuilder.java

Reply:
I understand it can be hard. The question is: _Do we need to solve this issue in DD?_
If some software/API is providing a field with the name _bug_url_, I suppose it should be a valid URL every time.
Is it possible to open an issue/question in _Bugcrowd_: _Why are they sending non-URL information in the field called `bug_url`?_

Reply:
I agree, it's just that I want to make a powerful extractor for this integration, because of the human-nature of it. It won't produce a good amount of data otherwise and I'd think it's a shame :/ I'm even considering to try and extract URLs from the submission title, because I see in our internal findings that most researchers don't bother submitting properly...

Reply:
You're right, I give up :P I think some submissions have disgusting inputs... I won't parse the description to find URLs either, but I'm still getting errors from the db column... `django.db.utils.DataError: (1406, "Data too long for column 'protocol' at row 1")`

Reply:
I just saw, we have people submitting bug_urls like a csv too, like `url1, url2, url3`

Reply:
As I remember correctly, the length of the field is not checked in `Endpoint.from_uri(...)` but in `Endpoint.save()`. So you can try to call:
```python
endpoint = Endpoint.from_uri(bug_url)
endpoint.protocol = endpoint.protocol[:20]
```

Reply:
But keep in mind, it will cut everything over 20 so 
```
abcdeabcdeabcdeabcdex://host/
```
and 
```
abcdeabcdeabcdeabcdey://host/
```
will be merged to `abcdeabcdeabcdeabcde://host/`. Not so sure it is the best solution. I prefer to solve it in a different way.

Reply:
Yeah, I dumped the protocols I find when parsing all the submissions we have.... and here's a clear bug :P 
The quick fix by cutting to 20 chars works, but looks at this:
Someone submitted `aaa-bbbbbb.cccccc.dddddd.eeeeee.com:443` as a bug_url
The protocol is weirdly parsed, and basically, the path becomes 443, and the protocol becomes `aaaa-bbbbb.ccccc`

Reply:
that's a bug on https://pypi.org/project/hyperlink/ https://github.com/python-hyper/hyperlink/issues/170

Reply:
Oooo, I see. I highly recommend to use this workaround https://github.com/DefectDojo/django-DefectDojo/blob/master/dojo/tools/generic/parser.py#L61-L66

Reply:
`aa.bb.cc:443` is not a valid URI/URL. But this ^ workaround works well.

Reply:
Alright I did that, it kinda works, the hosts are still broken because some just have "API Call" in the hostname, but it's fine I guess for now...